imsim/skycat.py

```diff
@@ -0,0 +1,374 @@
+"""
+Interface to obtain objects from skyCatalogs.
+"""
+import math
+import numpy as np
+from galsim.config import InputLoader, RegisterInputType, RegisterValueType, \
+    RegisterObjectType, RegisterSEDType
+from galsim import CelestialCoord
+import galsim
+from desc.skycatalogs import skyCatalogs
+from .instcat import get_radec_limits
+
+
+class SkyCatalogInterface:
+    """Interface to skyCatalogs package."""
+    _bp500 = galsim.Bandpass(galsim.LookupTable([499,500,501],[0,1,0]),
+                             wave_type='nm').withZeropoint('AB')
+
+    # Using area-weighted effective aperture over FOV
+    # from https://confluence.lsstcorp.org/display/LKB/LSST+Key+Numbers
+    _rubin_area = 0.25 * np.pi * 649**2  # cm^2
+
+    def __init__(self, file_name, wcs, obj_types=None, edge_pix=100,
+                 flip_g2=True, logger=None):
+        """
+        Parameters
+        ----------
+        file_name : str
+            Name of skyCatalogs yaml config file.
+        wcs : galsim.WCS
+            WCS of the image to render.
+        obj_types : list-like [None]
+            List or tuple of object types to render, e.g., ('star', 'galaxy').
+            If None, then consider all object types.
+        edge_pix : float [100]
+            Size in pixels of the buffer region around nominal image
+            to consider objects.
+        flip_g2 : bool [True]
+            Flag to flip sign of g2.
+        logger : logging.Logger
+            Logger object.
+        """
+        logger = galsim.config.LoggerWrapper(logger)
+        if obj_types is not None:
+            logger.warning(f'Object types restricted to {obj_types}')
+        self.file_name = file_name
+        self.wcs = wcs
+        self.flip_g2 = flip_g2
+        sky_cat = skyCatalogs.open_catalog(file_name)
+        region = skyCatalogs.Box(*get_radec_limits(wcs, logger, edge_pix)[:4])
+        self.objects = sky_cat.get_objects_by_region(region,
+                                                     obj_type_set=obj_types)
+        self._index_subcomponents()
+
+    def _index_subcomponents(self):
+        """
+        Create a list of all object subcompoents so that galsim.config
+        can use indexes corresponding the GSObjects, in contrast to
+        skyCatalog objects, which can correpond to composites of
+        GSObjects.
```

Comment:
Why do we want to do this?  I thought we were going to keep the subcomponents together as a single `galsim.Sum` object.  (Or `galsim.ChromaticSum` when chromatic=True.)

Reply:
Simply ignorance on my part.  I had no idea that `galsim.Sum` is an option.